File: app/src/main/java/de/markusdamm/pokerapp/data/PlayerStatistic.java
package de.markusdamm.pokerapp.data;

import android.util.Pair;

import java.util.ArrayList;
import java.util.List;

import de.markusdamm.pokerapp.utils.Utils;

/**
 * Created by Markus Damm on 30.03.2015.
 */
public class PlayerStatistic implements Comparable<PlayerStatistic> {

    private int bestPlace;
    private int worstPlace;
    private int wins;
    private int headUps;
    private int podiums;
    private int lastPlaces;
    private int participations;
    private int minuits;
    private int beatenPlayers;
    private int participators;
    private int sumOfPlaces;
    private int multikills;
    private int mostKills;
    private double median;
    private double sd;
    private double normalizedMean;
    private double average;
    private List<String> killed = new ArrayList<>();
    private int mostDeaths;
    private List<String> killers = new ArrayList<>();
    private Player player;
    private ArrayList<String> strings;

    private String value1, value2, value3;

    public final static String stBestPlace = "Beste Platzierung";
    public final static String stWorstPlace = "Schlechteste Platzierung";
    public final static String stWins = "Siege";
    public final static String stHeadUps = "Heads-Ups";
    public final static String stPodiums = "Podiumsplätze";
    public final static String stParticipations = "Anzahl an Teilnahmen";
    public final static String stMinuits = "Gespielte Zeit";
    public final static String stBeatenPlayers = "Anzahl rausgeworfener Spieler";
    public final static String stNumberOfOponents = "Anzahl Gegner";
    public final static String stSumOfPlaces = "Summe der Plätze";
    public final static String stWorsePlayer = "Anzahl schlechtere Spieler bei Teilnahme";
    public final static String stAveragePlace = "Durchschnittliche Platzierung";
    public final static String stMultikills = "Multikills";
    public final static String stLastPlaces = "Letzte Plätze";
    public final static String stMostKills = "Häufigste getötete Gegner";
    public final static String stMostDeaths = "Am häufigsten getötet von";
    public final static String stSD = "Standardabweichung";
    public final static String stMedian = "Median";
    public final static String stNormalizedMean = "Normalisierter Durchschnitt über Teilnehmerzahl";


    public PlayerStatistic(Player player){
        this.player = player;
        strings = new ArrayList<>();
        fillStrings();
    }

    public int getSumOfPlaces(){
        return sumOfPlaces;
    }

    public int getWorstPlace() {
        return worstPlace;
    }

    public int getBestPlace() {
        return bestPlace;
    }

    public int getWins() {
        return wins;
    }

    public int getParticipations() {
        return participations;
    }

    public int getMinuits() {
        return minuits;
    }

    public int getBeatenPlayers() {
        return beatenPlayers;
    }

    public int getParticipators() {
        return participators;
    }

    public Player getPlayer() {
        return player;
    }

    public int getMultikills() {
        return multikills;
    }


    public int getPodiums() {
        return podiums;
    }

    public int getHeadUps() {
        return headUps;
    }

    public int getLastPlaces() {
        return lastPlaces;
    }

    public int getMostKills() {
        return mostKills;
    }

    public int getMostDeaths() {
        return mostDeaths;
    }

    public String getMostKilled() {
        String ret = "";
        for (String kill: killed) {
            if (ret == "") {
                ret = kill;
            } else {
                ret = ret + ", " + kill;
            }
        }
        return ret;
    }

    public String getMostKillers() {
        String ret = "";
        for (String kill: killers) {
            if (ret == "") {
                ret = kill;
            } else {
                ret = ret + ", " + kill;
            }
        }
        return ret;
    }

    public double getMedian() {
        return median;
    }

    public double getSd() {
        return sd;
    }

    public double getAverage() {
        return average;
    }

    public double getNormalizedMean() {
        return normalizedMean;
    }


    public void setSumOfPlaces(int sumOfPlaces){
        this.sumOfPlaces = sumOfPlaces;
    }

    public void setBestPlace(int bestPlace) {
        this.bestPlace = bestPlace;
    }

    public void setWins(int wins) {
        this.wins = wins;
    }

    public void setParticipations(int participations) {
        this.participations = participations;
    }

    public void setMinuits(int minuits) {
        this.minuits = minuits;
    }

    public void setWorstPlace(int worstPlace) {
        this.worstPlace = worstPlace;
    }

    public void setBeatenPlayers(int beatenPlayers) {
        this.beatenPlayers = beatenPlayers;
    }

    public void setParticipators(int participators) {
        this.participators = participators;
    }

    public void setMultikills(int multikills) {
        this.multikills = multikills;
    }

    public void setPodiums(int podiums) {
        this.podiums = podiums;
    }

    public void setHeadUps(int headUps) {
        this.headUps = headUps;
    }

    public void setLastPlaces(int lastPlaces) {
        this.lastPlaces = lastPlaces;
    }


    public void setMostKills(Pair<Integer, List<String>> value) {
        this.mostKills = value.first;
        this.killers = value.second;
    }

    public void setMostDeaths(Pair<Integer, List<String>> value) {
        this.mostDeaths = value.first;
        this.killed = value.second;
    }


    public void setValues(String value1, String value2, String value3){
        this.value1 = value1;
        this.value2 = value2;
        this.value3 = value3;
    }

    private String buildMostKillers() {
        if (mostKills == -1) {
            return stMostKills + ": niemand";
        }

        return stMostKills + ": " + mostKills + " Mal: " + getMostKillers();
    }

    private String buildMostKilled() {
        if (mostDeaths == -1) {
            return stMostDeaths + ": niemanden";
        }

        return stMostDeaths + ": " + mostDeaths + " Mal von " + getMostKilled();
    }

    public void setMedian(double median) {
        this.median = median;
    }

    public void setSd(double sd) {
        this.sd = sd;
    }

    public void setNormalizedMean(double normalizedMean) {
        this.normalizedMean = normalizedMean;
    }

    public void setAverage(double average) {
        this.average = average;
    }


    public ArrayList<String> getStatisticList(){
        ArrayList<String> statisticList = new ArrayList<>();
        statisticList.add(stBestPlace + ": " + bestPlace);
        statisticList.add(stWorstPlace + ": " + worstPlace);
        statisticList.add(stWins + ": " + wins);
        statisticList.add(stHeadUps + ": " + headUps);
        statisticList.add(stPodiums + ": " + podiums);
        statisticList.add(stLastPlaces + ": " + lastPlaces);
        statisticList.add(stParticipations + ": " + participations);
        statisticList.add(stMinuits + ": " + minuits + " Minuten bzw. " + Utils.formatTimeToString(minuits));
        statisticList.add(stBeatenPlayers + ": " + beatenPlayers);
        statisticList.add(stNumberOfOponents + ": " + participators);
        statisticList.add(stSumOfPlaces + ": " + sumOfPlaces);
        statisticList.add(stWorsePlayer + ": " + Integer.toString(participators - sumOfPlaces));
        statisticList.add(stAveragePlace + ": " + average);
        statisticList.add(stNormalizedMean + ": " + normalizedMean);
        statisticList.add(stMedian + ": " + median);
        statisticList.add(stSD + ": " + sd);
        statisticList.add(stMultikills + ": " + multikills);
        statisticList.add(buildMostKillers());
        statisticList.add(buildMostKilled());

        return statisticList;
    }



    public <T extends Number> T getValue(String value){
        switch (value){
            case stBestPlace:
                return (T) Integer.valueOf(getBestPlace());
            case stWins:
                return (T) Integer.valueOf(getWins());
            case stMinuits:
                return (T) Integer.valueOf(getMinuits());
            case stBeatenPlayers:
                return (T) Integer.valueOf(getBeatenPlayers());
            case stNumberOfOponents:
                return (T) Integer.valueOf(getParticipators());
            case stSumOfPlaces:
                return (T) Integer.valueOf(getSumOfPlaces());
            case stWorsePlayer:
                return (T) Integer.valueOf(getParticipators() - getSumOfPlaces());
            case stParticipations:
                return (T) Integer.valueOf(getParticipations());
            case stAveragePlace:
                return (T) Double.valueOf(getAverage());
            case stMultikills:
                return (T) Integer.valueOf(getMultikills());
            case stHeadUps:
                return (T) Integer.valueOf(getHeadUps());
            case stPodiums:
                return (T) Integer.valueOf(getPodiums());
            case stWorstPlace:
                return (T) Integer.valueOf(getWorstPlace());
            case stLastPlaces:
                return (T) Integer.valueOf(getLastPlaces());
            case stMostKills:
                return (T) Integer.valueOf(getMostKills());
            case stMostDeaths:
                return (T) Integer.valueOf(getMostDeaths());
            case stMedian:
                return (T) Double.valueOf(getMedian());
            case stSD:
                return (T) Double.valueOf(getSd());
            case stNormalizedMean:
                return (T) Double.valueOf(getNormalizedMean());
            default:
                return (T) Integer.valueOf(-1);
        }
    }


    public void fillStrings(){
        strings.clear();
        strings.add(stBestPlace);
        strings.add(stWorstPlace);
        strings.add(stWins);
        strings.add(stHeadUps);
        strings.add(stPodiums);
        strings.add(stLastPlaces);
        strings.add(stParticipations);
        strings.add(stMinuits);
        strings.add(stBeatenPlayers);
        strings.add(stNumberOfOponents);
        strings.add(stSumOfPlaces);
        strings.add(stWorsePlayer);
        strings.add(stAveragePlace);
        strings.add(stMedian);
        strings.add(stSD);
        strings.add(stNormalizedMean);
        strings.add(stMultikills);
    }

    public ArrayList<String> getStrings(){
        return strings;
    }

    public List<String> inverseToSort() {
        List<String> ret = new ArrayList<>();
        ret.add(stBestPlace);
        ret.add(stSumOfPlaces);
        ret.add(stWorstPlace);
        ret.add(stLastPlaces);
        ret.add(stSD);
        ret.add(stMedian);
        ret.add(stAveragePlace);
        ret.add(stNormalizedMean);

        return ret;
    }

    public List<String> floatingValues() {
        List<String> ret = new ArrayList<>();
        ret.add(stSD);
        ret.add(stMedian);
        ret.add(stAveragePlace);
        ret.add(stNormalizedMean);
        return ret;
    }

    private int compareNumbers(Double x, Double y, String value) {
        if (inverseToSort().contains(value)) {
            return x.compareTo(y);
        } else {
            return y.compareTo(x);
        }
    }

    @Override
    public int compareTo(PlayerStatistic another) {
        Double x = this.getValue(value1).doubleValue();
        Double y = another.getValue(value1).doubleValue();

        if (x.doubleValue() != y.doubleValue()) {
            return compareNumbers(x, y, value1);
        }

        x = this.getValue(value2).doubleValue();
        y = another.getValue(value2).doubleValue();

        if (x.doubleValue() != y.doubleValue()) {
            return compareNumbers(x, y, value2);
        }

        x = this.getValue(value3).doubleValue();
        y = another.getValue(value3).doubleValue();

        return compareNumbers(x, y, value3);
    }
}

File: app/src/main/java/de/markusdamm/pokerapp/data/PlayerStatistic.kt
package de.markusdamm.pokerapp.data

import android.util.Pair

import java.util.ArrayList

import de.markusdamm.pokerapp.utils.Utils

/**
 * Created by Markus Damm on 30.03.2015.
 */
class PlayerStatistic(val player: Player?) : Comparable<PlayerStatistic> {

    var bestPlace: Int = 0
    var worstPlace: Int = 0
    var wins: Int = 0
    var headUps: Int = 0
    var podiums: Int = 0
    var lastPlaces: Int = 0
    var participations: Int = 0
    var minuits: Int = 0
    var beatenPlayers: Int = 0
    var participators: Int = 0
    var sumOfPlaces: Int = 0
    var multikills: Int = 0
    var mostKills: Int = 0
        private set
    var median: Double = 0.toDouble()
    var sd: Double = 0.toDouble()
    var normalizedMean: Double = 0.toDouble()
    var average: Double = 0.toDouble()
    private var killed: List<String> = ArrayList()
    var mostDeaths: Int = 0
        private set
    private var killers: List<String> = ArrayList()
    val strings: ArrayList<String> = ArrayList()

    private var value1: String? = null
    private var value2: String? = null
    private var value3: String? = null

    private val mostKilled: String
        get() {
            var ret = ""
            for (kill in killed) {
                ret = if (ret === "") {
                    kill
                } else {
                    "$ret, $kill"
                }
            }
            return ret
        }

    private val mostKillers: String
        get() {
            var ret = ""
            for (kill in killers) {
                ret = if (ret === "") {
                    kill
                } else {
                    "$ret, $kill"
                }
            }
            return ret
        }


    val statisticList: ArrayList<String>
        get() {
            return arrayListOf("$stBestPlace: $bestPlace",
                    "$stWorstPlace: $worstPlace",
                    "$stWins: $wins",
                    "$stHeadUps: $headUps",
                    "$stPodiums: $podiums",
                    "$stLastPlaces: $lastPlaces",
                    "$stParticipations: $participations",
                    "$stMinuits: $minuits  Minuten bzw. ${Utils.formatTimeToString(minuits)}",
                    "$stBeatenPlayers: $beatenPlayers",
                    "$stNumberOfOponents: $participators",
                    "$stSumOfPlaces: $sumOfPlaces",
                    "$stWorsePlayer : ${(participators - sumOfPlaces)}",
                    "$stAveragePlace: $average",
                    "$stNormalizedMean: $normalizedMean",
                    "$stMedian: $median",
                    "$stSD: $sd",
                    "$stMultikills: $multikills",
                    buildMostKillers(),
                    buildMostKilled())
        }

    init {
        fillStrings()
    }


    fun setMostKills(value: Pair<Int, List<String>>) {
        this.mostKills = value.first
        this.killers = value.second
    }

    fun setMostDeaths(value: Pair<Int, List<String>>) {
        this.mostDeaths = value.first
        this.killed = value.second
    }


    fun setValues(value1: String, value2: String, value3: String) {
        this.value1 = value1
        this.value2 = value2
        this.value3 = value3
    }

    private fun buildMostKillers(): String {
        return if (mostKills == -1) {
            "$stMostKills: niemand"
        } else "$stMostKills: $mostKills Mal: $mostKillers"

    }

    private fun buildMostKilled(): String {
        return if (mostDeaths == -1) {
            "$stMostDeaths: niemanden"
        } else "$stMostDeaths: $mostDeaths Mal von $mostKilled"

    }

    fun <T : Number> getValue(value: String): T {
        when (value) {
            stBestPlace -> return Integer.valueOf(bestPlace) as T
            stWins -> return Integer.valueOf(wins) as T
            stMinuits -> return Integer.valueOf(minuits) as T
            stBeatenPlayers -> return Integer.valueOf(beatenPlayers) as T
            stNumberOfOponents -> return Integer.valueOf(participators) as T
            stSumOfPlaces -> return Integer.valueOf(sumOfPlaces) as T
            stWorsePlayer -> return Integer.valueOf(participators - sumOfPlaces) as T
            stParticipations -> return Integer.valueOf(participations) as T
            stAveragePlace -> return java.lang.Double.valueOf(average) as T
            stMultikills -> return Integer.valueOf(multikills) as T
            stHeadUps -> return Integer.valueOf(headUps) as T
            stPodiums -> return Integer.valueOf(podiums) as T
            stWorstPlace -> return Integer.valueOf(worstPlace) as T
            stLastPlaces -> return Integer.valueOf(lastPlaces) as T
            stMostKills -> return Integer.valueOf(mostKills) as T
            stMostDeaths -> return Integer.valueOf(mostDeaths) as T
            stMedian -> return java.lang.Double.valueOf(median) as T
            stSD -> return java.lang.Double.valueOf(sd) as T
            stNormalizedMean -> return java.lang.Double.valueOf(normalizedMean) as T
            else -> return Integer.valueOf(-1) as T
        }
    }


    private fun fillStrings() {
        strings.clear()
        strings.addAll(listOf(stBestPlace, stWorstPlace, stWins, stHeadUps, stPodiums, stLastPlaces, stParticipations,
                stMinuits, stBeatenPlayers, stNumberOfOponents, stSumOfPlaces, stWorsePlayer, stAveragePlace, stMedian,
                stSD, stNormalizedMean, stMultikills))
    }

    private fun inverseToSort(): List<String> {
        return arrayListOf(stBestPlace, stSumOfPlaces, stWorstPlace, stLastPlaces, stSD, stMedian, stAveragePlace, stNormalizedMean)
    }

    fun floatingValues(): List<String> {
        return arrayListOf(stSD, stMedian, stAveragePlace, stNormalizedMean)
    }

    private fun compareNumbers(x: Double, y: Double, value: String?): Int {
        return if (inverseToSort().contains(value)) {
            x.compareTo(y)
        } else {
            y.compareTo(x)
        }
    }

    override fun compareTo(other: PlayerStatistic): Int {
        var x: Double? = this.getValue<Number>(value1!!).toDouble()
        var y: Double? = other.getValue<Number>(value1!!).toDouble()

        if (x!!.toDouble() != y!!.toDouble()) {
            return compareNumbers(x, y, value1)
        }

        x = this.getValue<Number>(value2!!).toDouble()
        y = other.getValue<Number>(value2!!).toDouble()

        if (x.toDouble() != y.toDouble()) {
            return compareNumbers(x, y, value2)
        }

        x = this.getValue<Number>(value3!!).toDouble()
        y = other.getValue<Number>(value3!!).toDouble()

        return compareNumbers(x, y, value3)
    }

    companion object {

        const val stBestPlace = "Beste Platzierung"
        const val stWorstPlace = "Schlechteste Platzierung"
        const val stWins = "Siege"
        const val stHeadUps = "Heads-Ups"
        const val stPodiums = "Podiumsplätze"
        const val stParticipations = "Anzahl an Teilnahmen"
        const val stMinuits = "Gespielte Zeit"
        const val stBeatenPlayers = "Anzahl rausgeworfener Spieler"
        const val stNumberOfOponents = "Anzahl Gegner"
        const val stSumOfPlaces = "Summe der Plätze"
        const val stWorsePlayer = "Anzahl schlechtere Spieler bei Teilnahme"
        const val stAveragePlace = "Durchschnittliche Platzierung"
        const val stMultikills = "Multikills"
        const val stLastPlaces = "Letzte Plätze"
        const val stMostKills = "Häufigste getötete Gegner"
        const val stMostDeaths = "Am häufigsten getötet von"
        const val stSD = "Standardabweichung"
        const val stMedian = "Median"
        const val stNormalizedMean = "Normalisierter Durchschnitt über Teilnehmerzahl"
    }
}
